feat(wxcc): agent-multi-login-changes

pagour98 opened the pull request:
<!--
Hey there,\
Thank you for taking the time to improve our code! 🙂\
Please let us know why this change is necessary and what testing you have performed. \
This ensures our reviewers understand the impact of your change. \

**IMPORTANT**
FAILING TO FILL OUT THIS TEMPLATE WILL RESULT IN REJECTION OF YOUR PULL REQUEST
This is for compliance purposes with FedRAMP program.
-->

# COMPLETES #<[SPARK-601180](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-601180)>

## This pull request addresses

* This PR enables agent multi-login alert in CC Widgets by making the below changes in CC SDK.

## by making the following changes

* In config.js file allowMultiLogin boolean flag is set to false, because we don't have to allow multi login similar to CC Desktop.
* AGENT_MULTI_LOGIN event has been emitted from SDK side, so that widgets can listen to it.

### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

Tested the agent multi login Websocket event by logging in agent multiple times.
Tested the other functionalities to ensure nothing breaks.
Tested SDK by following the CC SDK test plan suite 

Vidcast for the SDK Changes Testing:- https://app.vidcast.io/share/5ad06894-3ead-40db-8441-e4739e4569d3

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

Commit:
feat(wxcc): agent-multi-login

Commit:
feat(wxcc): added-test-cases

Commit:
feat(wxcc): fix-allow-multi-login-test-case

Commit:
Merge branch 'wxcc' of github.com:webex/webex-js-sdk into agent-multi-login

Commit:
feat(wxcc): fix-allow-multi-login-test-case

Commit:
feat(wxcc): remove-local-debug-stuff

pagour98 commented on packages/@webex/plugin-cc/src/config.ts:
```diff
@@ -2,7 +2,7 @@ import {LOGGER} from '@webex/calling';
 export default {
   cc: {
-    allowMultiLogin: true,
+    allowMultiLogin: false,
```
I have tried to set allowMultiLogin to false in webexConfig object inside CC Widgets but that we not working, So I think we need to set allowMultiLogin to false from SDK side

rsarika replied:
It should work. Approving the PR as still we can debug/test this with false also.

mkesavan13 replied:
Is there a TODO to fix this?

mkesavan13 replied:
Also, why do we need this change?

pagour98 replied:
@mkesavan13 , We need this because to tell backend that there should not be multi-login by agent and if there is a multi-login then send AGENT_MULTI_LOGIN event to the client, so that the client know that there is a multi-login.

This will not impact any existing functionality.

mkesavan13 reviewed (changes requested):
Please add the tests enumerated in the PR description as well as a demo vidcast

pagour98 commented:
> Please add the tests enumerated in the PR description as well as a demo vidcast

@mkesavan13 , Updated the PR description and added Vidcast for the same.

rarajes2 commented:
@pagour98  Let's update the SDK samples to show some notification on the UI for multi-login instead of showing in the console. Let's discuss 1-1 if help/clarification is needed.

Shreyas281299 committed:
Merge branch 'wxcc' into agent-multi-login

mkesavan13 reviewed (commented):
I'd like to know how the flow would look in the Widgets like, after this PR is merged. What SDK method do we call once they confirm? How does the other page get logged out?

Commit:
feat(cc-widget): added sample UI alert

Commit:
feat(cc-widget): latest pull and resolve merge conflicts

Commit:
feat(cc-widget): latest pull and resolve merge conflicts

pagour98 commented:
> @pagour98 Let's update the SDK samples to show some notification on the UI for multi-login instead of showing in the console. Let's discuss 1-1 if help/clarification is needed.

@rarajes2 , PR has been updated to show notification on samples page when there is agent multi login

pagour98 commented:
> I'd like to know how the flow would look in the Widgets like, after this PR is merged. What SDK method do we call once they confirm? How does the other page get logged out?

So when we set multilogin to false in sdk, then the server will send the AGENT_MULTI_LOGIN event. If the agent log in second time then the client will receive the event the above event and if the event is received on the existing session we show the alert dialog box and if the confirm is pressed we do web-socket registration again because in case of AGENT_MULTI_LOGIN event the server close down the existing web socket connection. When the agent sigin in new tab, agent-relogin method is called which silently login the agent in new session and the existing session will receive AGENT_MULTI_LOGIN EVENT

pagour98 commented:
> # COMPLETES #<[SPARK-601180](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-601180)>
> ## This pull request addresses
> * This PR enables agent multi-login alert in CC Widgets by making the below changes in CC SDK.
> 
> ## by making the following changes
> * In config.js file allowMultiLogin boolean flag is set to false, because we don't have to allow multi login similar to CC Desktop.
> * AGENT_MULTI_LOGIN event has been emitted from SDK side, so that widgets can listen to it.
> 
> ### Change Type
> * [ ]  Bug fix (non-breaking change which fixes an issue)
> * [x]  New feature (non-breaking change which adds functionality)
> * [ ]  Breaking change (fix or feature that would cause existing functionality to change)
> * [ ]  Documentation update
> * [ ]  Tooling change
> * [ ]  Internal code refactor
> 
> ## The following scenarios were tested
> < ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >
> 
> Tested the agent multi login Websocket event by logging in agent multiple times. Tested the other functionalities to ensure nothing breaks. Tested SDK by following the CC SDK test plan suite
> 
> Vidcast for the SDK Changes Testing:- https://app.vidcast.io/share/4a08ebd4-2540-4134-89d7-2279f1ec0c73
> 
> ### I certified that
> * [x]  I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
> * [x]  I discussed changes with code owners prior to submitting this pull request
> * [x]  I have not skipped any automated checks
> * [x]  All existing and new tests passed
> * [x]  I have updated the documentation accordingly
> 
> Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.
